presto-main/src/main/java/com/facebook/presto/metadata/MetadataManager.java

```diff
@@ -1210,6 +1214,28 @@ public void revokeTablePrivileges(Session session, QualifiedObjectName tableName
         return toListenableFuture(metadata.commitPageSinkAsync(connectorSession, tableHandle.getConnectorHandle(), fragments));
     }
 
+    @Override
+    public MetadataUpdates getMetadataUpdateResults(Session session, QueryManager queryManager, MetadataUpdates metadataUpdateRequests, QueryId queryId)
+    {
+        ConnectorId connectorId = metadataUpdateRequests.getConnectorId();
+        ConnectorMetadata metadata = getCatalogMetadata(session, connectorId).getMetadata();
+
+        if (queryManager != null && !queriesWithRegisteredCallbacks.contains(queryId)) {
```

Comment:
It's ok to remove multiple times as long as we can make sure they are cleaned up

Reply:
We want to add only 1 callback for a given queryId. If we don't check, we will end up adding callback everytime we call getMetadataUpdateResults. That might result is adding 10's of callbacks just for 1 query.